pecos/core/utils/clustering.hpp

```diff
@@ -31,6 +31,54 @@ enum {
     SKMEANS=5,
 }; /* partition_algo */
 
+enum {
+    CONSTANT_SAMPLE_SCHEDULE=0,
+    LINEAR_SAMPLE_SCHEDULE=1,
+}; /* sample strategies */
+
+extern "C" {
+    struct SampleSchedulerParam {
+        int strategy = CONSTANT_SAMPLE_SCHEDULE;
+        float sample_rate = 1.0;
+        float warmup_sample_rate, warmup_layer_rate;
+
+        SampleSchedulerParam(
+            int strategy,
+            float sample_rate,
+            float warmup_sample_rate,
+            float warmup_layer_rate
+        ): strategy(strategy),
+        sample_rate(sample_rate),
+        warmup_sample_rate(warmup_sample_rate),
+        warmup_layer_rate(warmup_layer_rate) {}
+    };
+} // end of extern C
+
+struct SampleScheduler {
+    // scheduler for sampling
+    SampleSchedulerParam param;
+    size_t warmup_layers;
+    size_t total_layers;
+
+    SampleScheduler(SampleSchedulerParam param, size_t total_layers): param(param), total_layers(total_layers) {
+        warmup_layers = size_t(total_layers * param.warmup_layer_rate);
+    }
+
+    float get_sample_rate(size_t layer) {
+        if(param.strategy == LINEAR_SAMPLE_SCHEDULE) {
+            return _get_linear_sample_rate(layer);
+        }
+        return param.sample_rate; // Constant strategy
+    }
+
+    float _get_linear_sample_rate(size_t layer) {
```

Comment:
need to comments to mathematically describe this sampling schedule.

Reply:
A comment describing the sampling schedule was added.